File: examples/python/part-6/19-mcp-security.py
# Illustrative examples for notes/part-6-multi-agent-security/19-mcp-security.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import os
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Protocol
from urllib.parse import urlparse


class RiskLevel(str, Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    CRITICAL = "Critical"


@dataclass
class MCPServer:
    id: str
    name: str
    version: str
    origin: str
    owner: str
    allowed_tools: List[str] = field(default_factory=list)
    allowed_domains: List[str] = field(default_factory=list)
    allowed_roots: List[str] = field(default_factory=list)
    risk: RiskLevel = RiskLevel.LOW
    reviewed: bool = False


@dataclass
class Registry:
    servers: Dict[str, MCPServer]

    def get(self, server_id: str) -> MCPServer:
        server = self.servers.get(server_id)
        if server is None:
            raise ValueError(f"mcp server is not allowlisted: {server_id}")
        if not server.reviewed:
            raise ValueError(f"mcp server is not reviewed: {server_id}")
        return server


@dataclass
class MCPToolCall:
    run_id: str
    user_id: str
    tenant_id: str
    server_id: str
    tool: str
    args: Dict[str, Any] = field(default_factory=dict)
    time: datetime = field(default_factory=datetime.utcnow)


def _contains(items: List[str], want: str) -> bool:
    return want in items


@dataclass
class Policy:
    registry: Registry

    def allow(self, call: MCPToolCall) -> MCPServer:
        server = self.registry.get(call.server_id)

        if not _contains(server.allowed_tools, call.tool):
            raise ValueError(
                f"tool {call.tool} is not allowed for server {call.server_id}"
            )

        if server.risk == RiskLevel.CRITICAL:
            raise RuntimeError("critical MCP server requires manual execution")

        return server


def validate_url(raw: str, allowed_domains: List[str]) -> None:
    parsed = urlparse(raw)
    if parsed.scheme != "https":
        raise ValueError("only https is allowed")

    host = (parsed.hostname or "").lower()
    for domain in allowed_domains:
        domain = domain.lower()
        if host == domain or host.endswith("." + domain):
            return

    raise ValueError(f"domain is not allowed: {host}")


def validate_path(path: str, allowed_roots: List[str]) -> None:
    clean = os.path.normpath(path)

    for root in allowed_roots:
        root_clean = os.path.normpath(root)
        try:
            rel = os.path.relpath(clean, root_clean)
        except ValueError:
            continue

        if rel == "." or (not rel.startswith("..") and not os.path.isabs(rel)):
            return

    raise ValueError(f"path is outside allowed roots: {path}")


class MCPClient(Protocol):
    def call_tool(
        self, server_id: str, tool: str, args: Dict[str, Any]
    ) -> Any: ...


class AuditLogger(Protocol):
    def log_mcp_call(
        self, call: MCPToolCall, decision: str, reason: str
    ) -> None: ...


@dataclass
class Executor:
    client: MCPClient
    policy: Policy
    audit: AuditLogger

    def call(self, call: MCPToolCall) -> Any:
        try:
            server = self.policy.allow(call)
        except Exception as exc:
            self.audit.log_mcp_call(call, "denied", str(exc))
            raise

        raw_url = call.args.get("url")
        if isinstance(raw_url, str):
            try:
                validate_url(raw_url, server.allowed_domains)
            except Exception as exc:
                self.audit.log_mcp_call(call, "denied", str(exc))
                raise

        raw_path = call.args.get("path")
        if isinstance(raw_path, str):
            try:
                validate_path(raw_path, server.allowed_roots)
            except Exception as exc:
                self.audit.log_mcp_call(call, "denied", str(exc))
                raise

        self.audit.log_mcp_call(call, "allowed", "policy passed")
        return self.client.call_tool(call.server_id, call.tool, call.args)

File: examples/typescript/part-6/19-mcp-security.ts
// Illustrative examples for notes/part-6-multi-agent-security/19-mcp-security.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

enum RiskLevel {
  Low = "Low",
  Medium = "Medium",
  High = "High",
  Critical = "Critical",
}

interface MCPServer {
  id: string;
  name: string;
  version: string;
  origin: string;
  owner: string;
  allowedTools: string[];
  allowedDomains: string[];
  allowedRoots: string[];
  risk: RiskLevel;
  reviewed: boolean;
}

class Registry {
  servers: Record<string, MCPServer>;

  constructor(servers: Record<string, MCPServer>) {
    this.servers = servers;
  }

  get(serverId: string): MCPServer {
    const server = this.servers[serverId];
    if (!server) {
      throw new Error(`mcp server is not allowlisted: ${serverId}`);
    }
    if (!server.reviewed) {
      throw new Error(`mcp server is not reviewed: ${serverId}`);
    }
    return server;
  }
}

interface MCPToolCall {
  runId: string;
  userId: string;
  tenantId: string;
  serverId: string;
  tool: string;
  args: Record<string, unknown>;
  time: Date;
}

function contains(items: string[], want: string): boolean {
  return items.includes(want);
}

class Policy {
  registry: Registry;

  constructor(registry: Registry) {
    this.registry = registry;
  }

  allow(call: MCPToolCall): MCPServer {
    const server = this.registry.get(call.serverId);

    if (!contains(server.allowedTools, call.tool)) {
      throw new Error(
        `tool ${call.tool} is not allowed for server ${call.serverId}`,
      );
    }

    if (server.risk === RiskLevel.Critical) {
      throw new Error("critical MCP server requires manual execution");
    }

    return server;
  }
}

function validateURL(raw: string, allowedDomains: string[]): void {
  let parsed: URL;
  try {
    parsed = new URL(raw);
  } catch (err) {
    throw err;
  }

  if (parsed.protocol !== "https:") {
    throw new Error("only https is allowed");
  }

  const host = parsed.hostname.toLowerCase();
  for (const domain of allowedDomains) {
    const normalized = domain.toLowerCase();
    if (host === normalized || host.endsWith(`.${normalized}`)) {
      return;
    }
  }

  throw new Error(`domain is not allowed: ${host}`);
}

// Illustrative path normalization (no node:path). Production: use a vetted library.
function normalizePath(path: string): string {
  const isAbsolute = path.startsWith("/") || /^[A-Za-z]:[\\/]/.test(path);
  const segments = path.replace(/\\/g, "/").split("/");
  const stack: string[] = [];

  for (const segment of segments) {
    if (segment === "" || segment === ".") {
      continue;
    }
    if (segment === "..") {
      if (stack.length > 0) {
        stack.pop();
      }
      continue;
    }
    stack.push(segment);
  }

  const joined = stack.join("/");
  if (isAbsolute) {
    return `/${joined}`;
  }
  return joined;
}

function isWithinRoot(clean: string, rootClean: string): boolean {
  if (clean === rootClean) {
    return true;
  }
  const prefix = rootClean.endsWith("/") ? rootClean : `${rootClean}/`;
  return clean.startsWith(prefix);
}

function validatePath(path: string, allowedRoots: string[]): void {
  const clean = normalizePath(path);

  for (const root of allowedRoots) {
    const rootClean = normalizePath(root);
    if (isWithinRoot(clean, rootClean)) {
      return;
    }
  }

  throw new Error(`path is outside allowed roots: ${path}`);
}

interface MCPClient {
  callTool(serverId: string, tool: string, args: Record<string, unknown>): unknown;
}

interface AuditLogger {
  logMCPCall(call: MCPToolCall, decision: string, reason: string): void;
}

class Executor {
  client: MCPClient;
  policy: Policy;
  audit: AuditLogger;

  constructor(client: MCPClient, policy: Policy, audit: AuditLogger) {
    this.client = client;
    this.policy = policy;
    this.audit = audit;
  }

  call(call: MCPToolCall): unknown {
    let server: MCPServer;
    try {
      server = this.policy.allow(call);
    } catch (err) {
      this.audit.logMCPCall(call, "denied", String(err));
      throw err;
    }

    const rawURL = call.args.url;
    if (typeof rawURL === "string") {
      try {
        validateURL(rawURL, server.allowedDomains);
      } catch (err) {
        this.audit.logMCPCall(call, "denied", String(err));
        throw err;
      }
    }

    const rawPath = call.args.path;
    if (typeof rawPath === "string") {
      try {
        validatePath(rawPath, server.allowedRoots);
      } catch (err) {
        this.audit.logMCPCall(call, "denied", String(err));
        throw err;
      }
    }

    this.audit.logMCPCall(call, "allowed", "policy passed");
    return this.client.callTool(call.serverId, call.tool, call.args);
  }
}

const CONTROL_PHRASES = [
  "ignore previous",
  "call tool",
  "run command",
  "execute",
  "system:",
];

function validateToolOutput(raw: string, maxLen: number): string {
  if (raw.length > maxLen) {
    throw new Error(`tool output exceeds max length: ${maxLen}`);
  }

  const lower = raw.toLowerCase();
  for (const phrase of CONTROL_PHRASES) {
    if (lower.includes(phrase)) {
      throw new Error(`tool output contains control instruction: ${JSON.stringify(phrase)}`);
    }
  }

  return raw;
}
